### PR TITLE
Http header error method

### PR DESCRIPTION
**Pull request by cproensa**
Replacing PR #660 

Fixes #0013048

### CODE REVIEW

File: config_defaults_inc.php [hunk 1]
```diff
@@ -4140,6 +4140,8 @@
  * - DISPLAY_ERROR_INLINE  Display a one line error and continue execution.
  * - DISPLAY_ERROR_NONE    Suppress the error (no display). This is the default
  *                         behavior for unspecified errors constants.
+ * - DISPLAY_ERROR_HEADERS Stop and return an HTTP error header. No output
```
Comment (vboctor) [inline config_defaults_inc.php]: Should we have a define or an API that when called at the beginning of a script page, it directs how the error is handled, without changing the valid values for this?  Having this option here means that users can use this globally.

File: core/error_api.php [hunk 2]
```diff
@@ -284,6 +284,12 @@ function error_handler( $p_type, $p_error, $p_file, $p_line, array $p_context )
 				$g_error_handled = true;
 				break;
 
+			case DISPLAY_ERROR_HEADERS:
+				# @TODO Have a mapping for mantis error codes to appropiate HTTP error codes
```
Comment (vboctor) [inline core/error_api.php]: type: appropriate

File: core/error_api.php [hunk 3]
```diff
@@ -284,6 +284,12 @@ function error_handler( $p_type, $p_error, $p_file, $p_line, array $p_context )
...
+				header( ' ', true, 500 );
+				exit(1);
+				break;
```
Comment (vboctor) [inline core/error_api.php]: Why is there a break here?

File: core/error_api.php [hunk 4]
```diff
@@ -548,3 +554,12 @@ function error_alternate_class() {
 		return 'class="row-2"';
 	}
 }
+
+/**
+ * Shorthand function to swith error output method to use HTTP headers
```
Comment (vboctor) [inline core/error_api.php]: typo: switch

File: core/error_api.php [hunk 5]
```diff
@@ -284,6 +284,12 @@ function error_handler( $p_type, $p_error, $p_file, $p_line, array $p_context )
 				$g_error_handled = true;
 				break;
 
+			case DISPLAY_ERROR_HEADERS:
+				# @TODO Have a mapping for mantis error codes to appropiate HTTP error codes
+				header( ' ', true, 500 );
```
Comment (vboctor) [inline core/error_api.php]: Shouldn't we return error_string( $p_error )?  Also for code, I would return the code for bad request (400) except in the case where the error is access denied, in which case I would return a forbidden (403) error code.

File: config_defaults_inc.php (hunk 1 above)
Comment (vboctor) [inline config_defaults_inc.php]: See install.php for example where it sets up expected behaviors for error handling, timeout, and maintenance mode.

    error_reporting( E_ALL );
    @set_time_limit( 0 );

    # Load the MantisDB core in maintenance mode. This mode will assume that
    # config/config_inc.php hasn't been specified. Thus the database will not be opened
    # and plugins will not be loaded.
    define( 'MANTIS_MAINTENANCE_MODE', true );

**Comment by cproensa** [comment]
> Should we have a define or an API that when called at the beginning of a script page, it directs how the error is handled, without changing the valid values for this? Having this option here means that users can use this globally.

That makes sense
i'd say a "supress output" mode. Where some behaviour is overrriden:
- Error handler: 
  - ignore error display methods that write HTML output
  - if error is type "halt", stop execution with an HTTP error header
- Log api:
  - ignore log display method for "screen"

switching this "supress error" mode may be through a define, or a api function (if it needs to change other options, termorary, or permanent)

Additionally, i'd add functionality to send error to the logging api, wether:
- having a log level for "errors"
- having an error display method "system log"

**Comment by vboctor** [comment]
The model makes sense.  I personally like the method approach over define.

Don't understand what you mean by "Additionally, i'd add functionality to send error to the logging api, wether: ...".

**Comment by cproensa** [comment]
> Don't understand what you mean by "Additionally, i'd add functionality to send error to the logging api,

Afaik, errors are not written into the log system, unless explicitly doing an "event_log" call
I think it is useful to have the errors logged too. So on the error handler, a event_log call could be used to log the detailed error, (besides the current display method used)
For controlling this kind of log messages, a new log level for "ERRORS", can be added.

And back to having a flag for "no html output" scripts, any error can be then registered by enabling the ERROR log level, and log destination to file.

**Comment by cproensa** [comment]
Actually, @dregad implemented already a define DISABLE_INLINE_ERROR_REPORTING
ab91f9eb84d7947bc7f318fa4adfa7c7ed4a1d1a

**Comment by cproensa** [comment]
Rewrited the PR using the already existant define: DISABLE_INLINE_ERROR_REPORTING

I'd move this constant to be a global variable, for this reason:
as explained by Damien, the usage is "by defining a constant prior to including core.php"
i find some cases where i want to disable error messages on ly through a specific part of code (im thinking of bug action group, for example). Seems like a define can't be used that way.

Also:
> Shouldn't we return error_string( $p_error )? Also for code, I would return the code for bad request (400) except in the case where the error is access denied, in which case I would return a forbidden (403)

I'm not sure if you can return any string as part of the header. I understand that the text associated to the error must be the standard one, that why i simply used the error code, with a dummy string. PHP fills that text by itself

**Comment by dregad** [comment]
I'm OK with the proposed approach to use a global variable rather than a define

**Comment by dregad** [comment]
@cproensa let me know if you want me to merge this as-is, and deal with the conversion from define to global as a separate issue/PR.

**Comment by cproensa** [comment]
I'm OK with this as is.
Thanks @dregad

**Comment by dregad** [comment]
Just realized there are merge conflicts. Could you kindly rebase on top of master to get rid of them ? Thanks.

**Comment by cproensa** [comment]
Ok, it's rebased now.
(I didn't get any conflicts though)

**Comment by dregad** [comment]
> (I didn't get any conflicts though)

Weird, Github complained that the merge could not be performed automatically... (I didn't actually try to merge)

**Pull request merged**